livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,408 @@
+        self._token_expires_at = time.time() + self._token_expires_in
+
+    async def _run(self, max_retry: int) -> None:
+        """
+        Run a single websocket connection to Assembly AI and make sure to reconnect
+        when something went wrong.
+        """
+        if self._token is None or time.time() > self._token_expires_at:
+            await self._get_temporary_token()
+
+        try:
+            retry_count = 0
+            while not self._closed:
+                try:
+                    live_config = {
+                        "sample_rate": self._sample_rate,
+                        "word_boost": self._opts.word_boost,
+                        "encoding": self._opts.encoding,
+                        "disable_partial_transcripts": self._opts.disable_partial_transcripts,
+                        "enable_extra_session_information": self._opts.enable_extra_session_information,
+                        "token": self._token,  # temporary authentication token
+                    }
+
+                    ws_url = "wss://api.assemblyai.com/v2/realtime/ws"
+                    url = f"{ws_url}?{urlencode(live_config).lower()}"
+                    ws = await self._session.ws_connect(url)
+                    retry_count = 0  # connected successfully, reset the retry_count
+
+                    await self._run_ws(ws)
+                except Exception:
+                    # Something went wrong, retry the connection
+                    if retry_count >= max_retry:
+                        print(
+                            f"failed to connect to Assembly AI after {max_retry} tries"
+                        )
+                        break
+
+                    retry_delay = min(retry_count * 2, 10)  # max 10s
+                    retry_count += 1  # increment after calculating the delay, the first retry should happen directly
+
+                    print(
+                        f"Assembly AI connection failed, retrying in {retry_delay}s",
+                    )
+                    await asyncio.sleep(retry_delay)
+        except Exception:
+            print("Assembly AI task failed")
+        finally:
+            self._event_queue.put_nowait(None)
+
+    async def _run_ws(self, ws: aiohttp.ClientWebSocketResponse) -> None:
+        """
+        This method can throw ws errors, these are handled inside the _run method
+        """
+
+        closing_ws = False
+        self._queue.put_nowait(self._END_UTTERANCE_SILENCE_THRESHOLD_MSG)
+
+        async def send_task():
+            nonlocal closing_ws
+            # forward inputs to Assembly AI
+            # if we receive a close message, signal it to Assembly AI and break.
+            # the recv task will then make sure to process the remaining audio and stop
+            while True:
+                data = await self._queue.get()
+                self._queue.task_done()
+
+                if isinstance(data, rtc.AudioFrame):
```

Comment:
worth adding an `else` case here to raise an Exception that there was an unexpected data type?